src/Microsoft.SqlTools.ServiceLayer/QueryExecution/SerializationService.cs

```diff
@@ -0,0 +1,256 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.SqlTools.Extensibility;
+using Microsoft.SqlTools.Hosting;
+using Microsoft.SqlTools.Hosting.Protocol;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.Contracts;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.DataStorage;
+using Microsoft.SqlTools.Utility;
+
+
+namespace Microsoft.SqlTools.ServiceLayer.QueryExecution
+{
+
+    [Export(typeof(IHostedService))]
+    public class SerializationService : HostedService<SerializationService>, IComposableService
+    {
+        private ConcurrentDictionary<string, DataSerializer> inProgressSerializations;
+
+        public SerializationService()
+        {
+            inProgressSerializations = new ConcurrentDictionary<string, DataSerializer>();
+        }
+
+        public override void InitializeService(IProtocolEndpoint serviceHost)
+        {
+            Logger.Write(TraceEventType.Verbose, "SerializationService initialized");
+            serviceHost.SetRequestHandler(SerializeDataRequest.Type, HandleSerializeDataRequest);
+        }
+
+        /// <summary>
+        /// Process request to save a resultSet to a file in CSV format
+        /// </summary>
+        internal async Task HandleSerializeDataRequest(SerializeDataRequestParams serializeParams,
+            RequestContext<SerializeDataResult> requestContext)
+        {
+            try
+            {
+                Validate.IsNotNull(nameof(serializeParams), serializeParams);
+                Validate.IsNotNullOrWhitespaceString("FilePath", serializeParams.FilePath);
+                DataSerializer serializer = null;
+                bool hasSerializer = inProgressSerializations.TryGetValue(serializeParams.FilePath, out serializer);
+                if (!hasSerializer)
+                {
```

Comment:
Is SqlToolsService actually multithreaded? If so this has a small chance for a race condition - if there isn't a serializer when we set hasSerializer but then another thread comes in and makes one before AddOrUpdate is called then we'll end up having two serializers for the same file. And generally this kind of check is an anti-pattern anyways.

I'd just create the new serializer and use TryAdd (ignoring if it already exists), simpler and avoids the possible race condition

Reply:
So my current thinking here is that the client will wait on the result from each call before sending more across. I believe given this, it should be fine.
I think to fix, we'd want to separate into 2 messages (start and update) and fail on start if another serializer existed. That would handle cases, for example, where folks clicked save twice. Let me take a look at this and think on it - this may be more robust against multiple requests.